File: src/main/java/dev/esophose/discordbot/command/arguments/EnumArgumentHandler.java
package dev.esophose.discordbot.command.arguments;

import dev.esophose.discordbot.command.DiscordCommandArgumentHandler;
import discord4j.core.object.entity.Guild;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import reactor.core.publisher.Mono;

public class EnumArgumentHandler<T extends Enum<T>> extends DiscordCommandArgumentHandler<T> {

    private Class<T> handledType;

    @Override
    protected Mono<T> handleInternal(Guild guild, String input) {
        Optional<T> match = Stream.of(this.getHandledType().getEnumConstants())
                .filter(x -> x.name().equalsIgnoreCase(input))
                .findFirst();

        return match.map(Mono::just).orElse(Mono.empty());
    }

    @Override
    public String getErrorMessage(Guild guild, String input) {
        return "Invalid " + this.handledType.getSimpleName() + " type [" + input + "]. Valid types: " +
                Stream.of(this.getHandledType().getEnumConstants()).map(x -> x.name().toLowerCase()).collect(Collectors.joining(", "));
    }

    @Override
    public Class<T> getHandledType() {
        return this.handledType;
    }

    /**
     * Must be called before running handleInternal() in order to parse the enum properly
     *
     * @param handledType The enum type to handle
     */
    public void setHandledType(Class<T> handledType) {
        this.handledType = handledType;
    }

}

File: src/main/kotlin/dev/esophose/discordbot/command/arguments/EnumArgumentHandler.kt
package dev.esophose.discordbot.command.arguments

import dev.esophose.discordbot.command.DiscordCommandArgumentHandler
import discord4j.core.`object`.entity.Guild
import reactor.core.publisher.Mono
import java.util.stream.Collectors
import java.util.stream.Stream
import kotlin.reflect.KClass

class EnumArgumentHandler<T : Enum<T>> : DiscordCommandArgumentHandler<T>() {

    /**
     * Must be set before running handleInternal() in order to parse the enum properly
     */
    var currentHandledType: KClass<*>? = null

    @Suppress("UNCHECKED_CAST")
    override val handledType: KClass<T>
        get() {
            return if (this.currentHandledType == null) {
                Enum::class as KClass<T>
            } else {
                currentHandledType as KClass<T>
            }
        }

    override fun handleInternal(guild: Guild, input: String): Mono<T> {
        val match = Stream.of(*this.handledType.java.enumConstants)
                .filter { x -> x.name.equals(input, ignoreCase = true) }
                .findFirst()

        return match.map { Mono.just(it) }.orElse(Mono.empty())
    }

    override fun getErrorMessage(guild: Guild, input: String): String {
        return "Invalid ${this.handledType.simpleName} type [$input]. Valid types: " +
                Stream.of(*this.handledType.java.enumConstants).map { it.name.toLowerCase() }.collect(Collectors.joining(", "))
    }

}
